WordPress/WordPressTest/PostCoordinatorTests.swift

```diff
@@ -56,16 +56,50 @@ class PostCoordinatorTests: XCTestCase {
 
         expect(post.remoteStatus).toEventually(equal(.pushing))
     }
+
+    func testAutoSavePostThatHasRemote() {
+        let postServiceMock = PostServiceMock()
+        let failedPostsFetcherMock = FailedPostsFetcherMock(context)
+        let postCoordinator = PostCoordinator(mainService: postServiceMock, backgroundService: postServiceMock, failedPostsFetcher: failedPostsFetcherMock)
+        let post = PostBuilder(context)
+            .withRemote()
+            .with(status: .draft)
+            .build()
+        failedPostsFetcherMock.postsAndActions = [post: .autoSave]
```

Comment:
About the first comment. I wasn't able to retrieve the post in `PostService` by doing something like this:

```
    func testAutoSavePostThatHasRemote() {
        let postServiceMock = PostServiceMock(managedObjectContext: context)
        let postCoordinator = PostCoordinator(mainService: postServiceMock, backgroundService: postServiceMock)
        let post = PostBuilder(context)
            .withRemote()
            .with(status: .draft)
            .with(remoteStatus: .failed)
            .build()

        postCoordinator.resume()

        expect(postServiceMock.didCallAutoSave).to(beTrue())
    }
```

If you have any idea how to make this work let me know. :)

About the 2nd: isn't that what this class is doing?

```
    private var context: NSManagedObjectContext!

    override func setUp() {
        super.setUp()
        context = TestContextManager().newDerivedContext()
    }

    override func tearDown() {
        super.tearDown()
        context = nil
    }
```

Reply:
I got it working by using:

- `toEventually()` since everything is happening in blocks
- `try! context.save()` because ¯\\\_(ツ)\_/¯ -- maybe because we are using different contexts somewhere and we have to make sure the `Post` is persisted? 

Here is the working diff: https://gist.github.com/shiki/12b2c3554d65c3580ba75adfd6691f6a

## TestContextManager

So this is how I understand `TestContextManager`. When it is initialized, it overrides `ContextManager.shared`:

https://github.com/wordpress-mobile/WordPress-iOS/blob/979540d22fa8d164f88d9f4917cc3031fedb1a2a/WordPress/WordPressTest/TestContextManager.m#L19-L29

Which means, after `setUp()`, all calls of `ContextManager.shared` will use that. 

And when it is deallocated, it removes the override:

https://github.com/wordpress-mobile/WordPress-iOS/blob/979540d22fa8d164f88d9f4917cc3031fedb1a2a/WordPress/WordPressTest/TestContextManager.m#L12-L17

If `setUp()` is like this: 

```swift 
        context = TestContextManager().newDerivedContext()
```

Doesn't that mean that `ContextManager.shared` is not overridden by `TestContextManager` anymore because it is deallocated right away? But maybe it's not deallocated? Who knows. ¯\\\_(ツ)\_/¯

Reply:
Fixed in 6a6354f3bcc480e668a35a8ec40da321041cd50b

About the `TestContextManager().newDerivedContext()` I debugged here and the behavior is the same as when we instantiate it separately. By doing that we'd save a few more lines. :)

Reply:
Interesting! Is it the `context` ivar that **retains** the `TestContextManager`?